internal/types/host/keepalived_host_collector_host.go

```diff
@@ -0,0 +1,270 @@
+
+	"github.com/cafebazaar/keepalived-exporter/internal/collector"
+	"github.com/cafebazaar/keepalived-exporter/internal/types/utils"
+)
+
+var sigNumSupportedVersion = version.Must(version.NewVersion("1.3.8"))
+
+// KeepalivedHostCollectorHost implements Collector for when Keepalived and Keepalived Exporter are both on a same host
+type KeepalivedHostCollectorHost struct {
+	pidPath string
+	version *version.Version
+	useJSON bool
+
+	SIGJSON  syscall.Signal
+	SIGDATA  syscall.Signal
+	SIGSTATS syscall.Signal
+}
+
+// NewKeepalivedHostCollectorHost is creating new instance of KeepalivedHostCollectorHost
+func NewKeepalivedHostCollectorHost(useJSON bool, pidPath string) *KeepalivedHostCollectorHost {
+	k := &KeepalivedHostCollectorHost{
+		useJSON: useJSON,
+		pidPath: pidPath,
+	}
+
+	var err error
+	k.version, err = k.getKeepalivedVersion()
+	if err != nil {
+		logrus.WithError(err).Warn("Version detection failed. Assuming it's the latest one.")
+	}
+
+	k.getSignals()
+
+	return k
+}
+
+func (k *KeepalivedHostCollectorHost) getSignals() {
+	if k.useJSON {
+		k.SIGJSON = k.sigNum("JSON")
+	}
+	k.SIGDATA = k.sigNum("DATA")
+	k.SIGSTATS = k.sigNum("STATS")
+}
+
+// GetKeepalivedVersion returns Keepalived version
+func (k *KeepalivedHostCollectorHost) getKeepalivedVersion() (*version.Version, error) {
+	cmd := exec.Command("bash", "-c", "keepalived -v")
+	var stdout, stderr bytes.Buffer
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	if err := cmd.Run(); err != nil {
+		logrus.WithFields(logrus.Fields{"stderr": stderr.String(), "stdout": stdout.String()}).WithError(err).Error("Error getting keepalived version")
+		return nil, errors.New("Error getting keepalived version")
+	}
+
+	version, err := utils.ParseVersion(stdout.String())
+	if err != nil {
+		logrus.WithError(err).Error("Failed to parse version output")
+		return nil, errors.New("Error parsing version output")
+	}
+
+	return version, nil
+}
+
+// Signal sends signal to Keepalived process
+func (k *KeepalivedHostCollectorHost) signal(signal os.Signal) error {
+	data, err := ioutil.ReadFile(k.pidPath)
+	if err != nil {
+		logrus.WithField("path", k.pidPath).WithError(err).Error("Can't find keepalived")
+		return err
+	}
+
+	pid, err := strconv.Atoi(strings.TrimSuffix(string(data), "\n"))
+	if err != nil {
+		logrus.WithField("path", k.pidPath).WithError(err).Error("Unknown pid found for keepalived")
+		return err
+	}
+
+	proc, err := os.FindProcess(pid)
+	if err != nil {
+		logrus.WithField("pid", pid).WithError(err).Error("Failed to find process")
+		return err
+	}
+
+	err = proc.Signal(signal)
+	if err != nil {
+		logrus.WithField("pid", pid).WithError(err).Error("Failed to send signal")
+		return err
+	}
+
+	// Wait 10ms for Keepalived to create its files
+	time.Sleep(10 * time.Millisecond)
+	return nil
+}
+
+// SigNum returns signal number for given signal name
+func (k *KeepalivedHostCollectorHost) sigNum(sigString string) syscall.Signal {
+	if !k.HasSigNumSupport() {
```

Comment:
I think this should be duplicated in all collectors, too :thinking:

Reply:
This could be extracted as a utils function.